public/app/features/query/state/updateQueries.ts

```diff
@@ -1,19 +1,49 @@
-import { DataQuery, DataSourceInstanceSettings, getDataSourceRef } from '@grafana/data';
+import { AppEvents, DataQuery, DataSourceApi, DataSourceInstanceSettings, getDataSourceRef } from '@grafana/data';
 import { isExpressionReference } from '@grafana/runtime/src/utils/DataSourceWithBackend';
+import appEvents from 'app/core/app_events';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { importDataSourcePlugin } from 'app/features/plugins/plugin_loader';
 
-export function updateQueries(
+export async function updateQueries(
   newSettings: DataSourceInstanceSettings,
   queries: DataQuery[],
   dsSettings?: DataSourceInstanceSettings
-): DataQuery[] {
+): Promise<DataQuery[]> {
   const datasource = getDataSourceRef(newSettings);
-
   // we are changing data source type
   if (dsSettings?.type !== newSettings.type) {
     // If changing to mixed do nothing
     if (newSettings.meta.mixed) {
       return queries;
     } else {
+      const dataSourceSrv = getDatasourceSrv();
+      let datasource: DataSourceApi;
+
+      try {
+        datasource = await dataSourceSrv.get(newSettings.uid);
+      } catch (error) {
+        datasource = await dataSourceSrv.get();
+      }
```

Comment:
maybe a bit cleaner to wrap this block with a function, then you can change the declaration to const